plugins/item_previews/web_client/views/ItemPreviewWidget.js

```diff
@@ -1,103 +1,175 @@
-        this.collection = new ItemCollection();
+        this._isSupportedItem = this._isSupportedItem.bind(this);
+        this.debouncedAddMoreItem = _.debounce(this.addMoreItem, 100);
 
-        this.collection.on('g:changed', function () {
-            this.trigger('g:changed');
-            this.render();
-        }, this).fetch({
-            folderId: settings.folderId
-        });
+        this.initialized = false;
+
+        this.addingMoreItems = false;
 
-        this.wrapPreviews = false;
+        this.supportedItems = [];
+    },
+
+    // Because the parent-child life cycle is not clean, this method is needed.
+    setCollection: function (collection) {
+        this.collection = collection;
+        this.collection.on('g:changed', () => {
+            this.supportedItems = this.collection.toJSON().filter(this._isSupportedItem);
+            if (this.supportedItems.length !== 0 &&
+                (this.isNotFull() || this.isNearBottom())) {
+                this.debouncedAddMoreItem();
+            }
+        });
     },
 
     _MAX_JSON_SIZE: 2e6 /* bytes */,
+    _LOAD_BATCH_SIZE: 2,
 
     _isSupportedItem: function (item) {
-        return this._isImageItem(item) || this._isJSONItem(item);
+        return this._isImageItem(item.name) || this._isJSONItem(item.name);
+    },
+
+    _isJSONItem: function (str) {
+        return /(json)$/i.test(str);
     },
 
-    _isJSONItem: function (item) {
-        return /\.(json)$/i.test(item.get('name'));
+    _isImageItem: function (str) {
+        return /(jpg|jpeg|png|gif)$/i.test(str);
     },
 
-    _isImageItem: function (item) {
-        return /\.(jpg|jpeg|png|gif)$/i.test(item.get('name'));
+    setElement: function ($el) {
+        this.initialized = false;
+        View.prototype.setElement.apply(this, arguments);
     },
 
     render: function () {
-        var supportedItems = this.collection.filter(this._isSupportedItem.bind(this));
-
-        var view = this;
-
-        view.$el.html(ItemPreviewWidgetTemplate({
-            items: supportedItems,
-            wrapPreviews: view.wrapPreviews,
-            hasMore: this.collection.hasNextPage(),
-            isImageItem: this._isImageItem,
-            isJSONItem: this._isJSONItem
-        }));
-
-        // Render any JSON files.
-        supportedItems.filter(this._isJSONItem).forEach(function (item) {
-            var id = item.get('_id');
-
-            // Don't process JSON files that are too big to preview.
-            var size = item.get('size');
-            if (size > this._MAX_JSON_SIZE) {
-                return view.$('.json[data-id="' + id + '"]').text('JSON too big to preview.');
-            }
+        if (!this.initialized) {
+            this.initialized = true;
+            this.renderedIndex = 0;
+            this.$el.html(ItemPreviewWidgetTemplate(this));
+            this.$('.g-widget-item-previews-container').on('scroll', (e) => this.onScroll(e));
+            this.debouncedAddMoreItem();
```

Comment:
I remember I found that the HierarchyWidget may call the ItemPreviewWidget too frequently or too early, causing duplicate calls so I have the debounce, but internal calls to addMoreItem are better managed.